### ISSUE TITLE
Revisit model-view transform support in bamboo.

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/bamboo#27]
This came up in a Google Group discussion, see https://groups.google.com/g/developing-interactive-simulations-in-html5/c/Srw0Yduz80Y. 

ModelViewTransform2 is the standard way that sims handle model-view coordinate transforms.  Unfortunately, bamboo currently uses ChartTransform instead of ModelViewTransform2.  This creates 3 significant problems:

(1) A sim may need to have both a ModelViewTransform2 and a ChartTransform, and they must be defined such that they result in the same transforms.
(2) ChartTransform supports only model-to-view transforms.  If (for example) you press on a chart, ChartTransform can't help you with transforming that press from view to model coordinates.
(3) This makes a complicated topic (coordinate transforms) even more complicated. 

I'm exactly sure how to proceed, will give it some thought.  Assigning to @samreid too, so we can discuss.

**Comment by samreid** [comment]
Bamboo currently supports nonlinear transforms, and I would not like to drop that feature by using ModelViewTransform2, which currently only supports linear transforms.  The other feature that bamboo handles nicely is support for setting the vertical transform independently from the horizontal transform. ModelViewTransform2 is mutable, but currently lacks support for setting one dimension at a time (though that feature could be added to ModelViewTransform2).

For the long term solution, I recommend creating an interface like `IModelViewTransform2` which has `ModelViewTransform2` and `ChartTransform` as concrete implementations.  Common code should only use the `IModelViewTransform2` interface.  We will have language-level support for interfaces if we proceed in https://github.com/phetsims/tasks/issues/987.  If we want to proceed before, that, we would have to fake it with a `ModelViewTransform2Def` "type" (not recommended).

**Comment by pixelzoom** [comment]
> Bamboo currently supports nonlinear transforms

Is this via ChartTransform's `xScale` and `yScale` options?  Does that really work for (e.g.) a log transform?   The implementation for `modelToView` looks like it scale value/min/max, but still does a linear interpolation:

```js
    return axisOrientation === Orientation.HORIZONTAL ?
           Util.linear( scale( modelRange.min ), scale( modelRange.max ), 0, viewDimension, scaledValue ) :
           Util.linear( scale( modelRange.max ), scale( modelRange.min ), 0, viewDimension, scaledValue );
```

**Comment by samreid** [comment]
I agree, "transform" may be a more natural term.  And they will need to be invertible so that bamboo can support view to model transforms.

**Comment by pixelzoom** [comment]
I agree that bamboo shouldn't use ModelViewTransform2.  And it also sounds like we're in agreement that the current capabilities of ChartTransform are lacking.  So I'm going to re-title this issue accordingly.

**Comment by samreid** [comment]
I've written an initial draft for this implementation and it is ready for review.  The demo looks like this:

![image](https://user-images.githubusercontent.com/679486/121710940-4e738c80-ca97-11eb-9289-6dd273b4862f.png)

@pixelzoom can you please review?

**Comment by samreid** [comment]
Some more improvements in the commits, including a fix for a problem CT identified.

**Comment by pixelzoom** [comment]
Before I review this...   https://github.com/phetsims/bamboo/issues/38 got me to thinking about an AxesNode with `doubleHead: false`, and wondering if it did the right thing -- the arrow head should point in the axis' positive direction. AxisNode implementation looks like it always assumes +x right, +y up.  And then I realized that ChartTransform is limited to +x right, +y up because it uses `Range` for `modelXRange` and `modelViewRange`.   This seems like a serious limitiation of a chart library. Should it be addressed?  @samreid thoughts?

**Comment by samreid** [comment]
Thanks, I moved the recommendation to a new issue.  It seems to be this issue could be reviewed independently before #39 is done, but you may save some time in reviewing both together once #39 is done.  Up to you if you want to review this as an intermediate step beforehand.

**Comment by pixelzoom** [comment]
Changes look OK, but a few things:

* Related methods were not grouped together in the source. I reorganized in https://github.com/phetsims/bamboo/commit/07b76a46708f75877bb8fee17f724a21b5605493.

* `viewToModelPoint` is a very different name than `modelToViewPosition`. ModelViewTransform2 uses the "position" terminology, so I went with that in https://github.com/phetsims/bamboo/commit/9a852e9949ef4a52b807502ca2ddea5abdad2fb9.

* Model-to-view API has many convenience methods that are missing for view-to-model. I noted them with a TODO in https://github.com/phetsims/bamboo/commit/3ecfd470e394f2c7c2c2ee3665039f2dc1678a15.

Back to @samreid.
